### DISCUSSION TITLE
Onewire DS18B20

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#852] [Ideas]
I included in my project the module "Onewire" and I added to my moddable two until 4 sensors DS18B20..
I have two problem:
1) When the cable of the sensors is very long (over 10 m) the module stops to work. This is because the wave form change so it's necessary to modify file owb_rmt.c. I changed  some value of slot duration like OW_DURATION_RESET , OW_DURATION_SLOT.. and now it works. It could be nice to have the possibility to modify them dynamically.
2) DS18B20 have a scratchpad with an Eeprom with 3 byte. The first two byte are used for threshold alarm and the last for the resolution.
In the module it seems that when this scratchpad is called the first byte is set to "78" (??) and the command "sensor.resolution = 9" no effect on the resolution.

**Comment by phoddie** [comment]
@wilberforce, can you help with this?

**Comment by wilberforce** [comment]
sure!

1. https://www.maximintegrated.com/en/design/technical-documents/tutorials/1/148.html
> OW_DURATION_RESET , OW_DURATION_SLOT.. and now it works. It could be nice to have the possibility to modify them dynamically.

You could set this up so that these values in the manifest the same way these are set up:

https://github.com/Moddable-OpenSource/moddable/blob/e9742b49d5132f2723d3a5fca45b65cf3818d01a/modules/drivers/onewire/esp/modonewire.c#L36-L38

Then you can override in your manifest and use the #defined `MODDEF_ONEWIRE_DURATION_RESET` and `MODDEF_ONEWIRE_DURATION_SLOT`

2. > sensor.resolution = 9" no effect on the resolution.

So when you read the resolution after setting it does not come back as 9?
I use resoultion 12 most of the time as its most accurate.

Does the example work for you with a short run?

https://github.com/Moddable-OpenSource/moddable/blob/e9742b49d5132f2723d3a5fca45b65cf3818d01a/examples/drivers/onewire/main.js#L63-L69

**Reply by phoddie** [reply]
(@wilberforce, thanks for sharing your expertise here!)

**Comment by mauroForlimpopoli** [comment]
Thank wilberforce, 
1) sorry for my inexperience, but I can inset "OW_DURATION_RESET" in my Manifest, but how can I change it dynamically in the main.js ?
2) The example about the change of the resolution doesn't work. I used DS18B20 bought "online" and some of them are "Fake", the sensors measure temperature, but the Eeprom memory is not present (Sure of that !) Unfortunately my sensor are incapsulate, so its about impossible to verify the originality  of the sensors Only in some of them I read the Eeprom. What I want to say is that  if you tested the module about the write on the scratchpad  and you haven't the problem maybe it depends from the sensors.

**Reply by wilberforce** [reply]
1. Sorry - I don't understand why do you want to change the `DURATION_RESET` dynamically in main.js? If you have values that are working why do they need to be adjusted?
The other thing you can do is try different resistors - what value are you using now?  Try a 10K or 3.3K and see if that works with the existing values

2. Did you want to change the resolution? As far as the alarm goes I think you are better off implementing it in javascript.

**Reply by mauroForlimpopoli** [reply]
Dear wilberforce, 
thank you for your answer,
1) Sometimes I add to the board a sensor with 10m. of cable , sometimes 2 sensors with 7m ...and so on. It seems that in each condition is preferible to tune the time of read. I used a 4.7k resistor. Now I reduce it to 2.2k and it goes better. 
2) When the command to change the resolution is sended, I suppose you need to write all the three byte of the scratchpad. Are the  bytes of the allarm written with other value or replace the old ones?
regards

**Reply by wilberforce** [reply]
1. You would need to update the driber to have the set and gets method  `duration_reset(480)` and duration_slot(75).
So you have changed the resistor value recently - does this still require the timings to be changed?

2. Are you using the alarm? The set could be set up to preseve the other scratch pad value - however you usually set the resolution at start up - then this would not effect the alarm.

**Reply by mauroForlimpopoli** [reply]
Dear wilberforce, 
1) I changed the resistor to 2,2k and now it goes better, but if the cable is longer I think to reduce the resistor until 1.5k. 
2) I'm not using alarm, but I'm using the 2 byte of Eeprom to store inside some information about the sensors
 It seems the the function "set scratchpad(bytes)" is used continuously  
I think that this function
```
 const bus = this.bus;
 bus.reset();
 bus.select(this.id);
 bus.write(0x4E); // Write scratchpad
 for (let i = 0, length = bytes.length; i < length; i++) {
    bus.write(bytes[i]);
 }
 bus.select(this.id);
bus.write(0x48); // Copy to scratch pad
bus.reset();
``` 
when the "bytes" is undefined send the" 0x4E" command // Write scratchpad
then no byte is wrote, 
but the command  bus.select(this.id); send the value "Match Rom [55H]"
this value is accepted like first byte to write in the scratchpad and so in my next read always I find the value of 55H like !° byte of EEprom

Thank you for your assistance

Regards

**Reply by wilberforce** [reply]
Sorry, I have not used this feature.

It might be easier to store the device config in javascript - using the hex code of the device as an index.

**Reply by mauroForlimpopoli** [reply]
I think that this simple modification can avoid to write over the previous value
```
    const bus = this.bus;
    bus.reset();
    bus.select(this.id);
	if (bytes.length > 0) {
		bus.write(0x4E); // Write scratchpad
		bus.write(bytes[0]);
		bus.write(bytes[1]);
		bus.write(bytes[2]);
		bus.select(this.id);
    	bus.write(0x48); // Copy to scratch pad
	}
    bus.reset();
```
Let me know if you are agree
regards

**Reply by wilberforce** [reply]
I'm sure how this is trying to prevent overwrite unless you where sending more than 4 bytes.

Perhaps:

`bytes.length > 0` should be `bytes.length == 3` to check there are 3 bytes to write.
Also just checking you are not doing this while a temperature request going on as they can take up to 750ms to complete.

**Comment by wilberforce** [comment]
I ran the example code with a couple of DS18b20 ... this is what I see:

```
C:\Users\rhys\Projects\moddable\modules\drivers\onewire\onewire.js (45) # Break: step!
284273f90400005a
,28a39f840200005b
Found first:284273f90400005a
Found 2nd:28a39f840200005b
Sensor present: true
NULL Sensor present: false
res Before:12
res - should be 9 :9
res - should be 12 :12
0 284273f90400005a: 27.875
1 28a39f840200005b: 27.75
Got temp!27.5625
```

So the resolution does appear to set using the examples.

**Reply by mauroForlimpopoli** [reply]
I suppose the issue is in my sensors. I will buy other ones and I will test it again
Thank you very much!
